src/AdoNet/Orleans.GrainDirectory.AdoNet/AdoNetGrainDirectory.cs

```diff
@@ -0,0 +1,199 @@
+                .LookupGrainActivationAsync(_clusterId, name, grainId.ToString())
+                .WaitAsync(lifetime.ApplicationStopping);
+
+            return entry?.ToGrainAddress();
+        }
+        catch (Exception ex)
+        {
+            LogFailedToLookup(ex, _clusterId, grainId);
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Registers a new grain activation.
+    /// </summary>
+    /// <param name="address">The grain address.</param>
+    /// <returns>The new or current grain address.</returns>
+    public async Task<GrainAddress> Register(GrainAddress address)
+    {
+        ArgumentNullException.ThrowIfNull(address);
+
+        try
+        {
+            var queries = await GetQueriesAsync();
+
+            // this call is expected to register a new entry or return the existing one if found in a thread safe manner
+            var entry = await queries
+                .RegisterGrainActivationAsync(_clusterId, name, address.GrainId.ToString(), address.SiloAddress.ToParsableString(), address.ActivationId.ToParsableString())
+                .WaitAsync(lifetime.ApplicationStopping);
+
+            LogRegistered(_clusterId, address.GrainId, address.SiloAddress, address.ActivationId);
+
+            return entry.ToGrainAddress();
+        }
+        catch (Exception ex)
+        {
+            LogFailedToRegister(ex, _clusterId, address.GrainId, address.SiloAddress, address.ActivationId);
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Unregister an existing grain activation.
+    /// </summary>
+    /// <param name="address">The grain address.</param>
+    public async Task Unregister(GrainAddress address)
+    {
+        ArgumentNullException.ThrowIfNull(address);
+
+        try
+        {
+            var queries = await GetQueriesAsync();
+
+            var count = await queries
+                .UnregisterGrainActivationAsync(_clusterId, name, address.GrainId.ToString(), address.ActivationId.ToParsableString())
+                .WaitAsync(lifetime.ApplicationStopping);
+
+            if (count > 0)
+            {
+                LogUnregistered(_clusterId, address.GrainId, address.SiloAddress, address.ActivationId);
+            }
+        }
+        catch (Exception ex)
+        {
+            LogFailedToUnregister(ex, _clusterId, address.GrainId, address.SiloAddress, address.ActivationId);
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Unregisters all grain activations in the specified set of silos.
+    /// </summary>
+    /// <param name="siloAddresses">The set of silos.</param>
+    public async Task UnregisterSilos(List<SiloAddress> siloAddresses)
+    {
+        ArgumentNullException.ThrowIfNull(siloAddresses);
+
+        if (siloAddresses.Count == 0)
+        {
+            return;
+        }
+
+        try
+        {
+            var siloAddressesAsString = string.Join('|', siloAddresses.Select(x => x.ToParsableString()));
```

Comment:
Move this below to L121.

This way it needs to "survive" the `await GetQueriesAsync`, so must be stored on the state machine box. When moved down, that box becomes smaller, so less work for the GC.